Turn Lapply case of Lambda.lambda into a record

nojb opened the pull request:
This is a follow-up to #111, making the arguments of the `Lapply` case of `Lambda.lambda` into a record, in analogy to the change in #133, which treats the case of `Lfunction`.

Drup commented:
Why the `ap_` prefix ? Record disambiguation should figure it out pretty much all the time since it's used almost only inside a pattern match on Lambda, so you can remove the prefix, which in turns mean you can use record field elision more often.

nojb commented:
Yes, I was in two minds about this one.  In general I try not to use record disambiguation for explicitness-sake.  But it is true that in this case it would work perfectly because it is always used inside an `Lapply`.  I will update the PR.

Drup commented:
Don't update just yet, because the core team might want to avoid disambiguation in the compiler code base, I don't know their stance on this. :)

nojb commented:
Yes, that was another concern - but #133 already forgoes prefixes in the `lfunction` record, so my guess is that it should not be a problem.

gasche commented:
I forgot about the prefixes when merging #133 (I may have to revisit this patch), but indeed the consensus is that it's better to keep them for now.

I dislike the use of optional arguments in `make_apply`. I am absolutely sure making `loc` optional is a mistake: when you make an argument optional you should assume that most call-sites will use the default value (because that's the path of least effort), and having no location by default is wrong. It needs to be mandatory so that people are forced to think about what the location should be. The argument for `tailcall` is less clear, but I would rather make everything non-optional.

nojb commented:
Agree on both counts. However, the `~tailcall` argument requires a label and I think that `~tailcall` is not a good name since `~tailcall:false` only means that the application does not have the `[@tailcall]` annotation.  Any ideas?

gasche commented:
`must_tail`? `tailcall_required`?

nojb commented:
I would go with `must_tail` as it is shorter.

nojb committed:
Turn Lapply case of Lambda.lambda into a record

nojb commented:
Both arguments `must_tail` and `loc` are now mandatory.

gasche commented:
We looked at this again just now. I would like the merge the change, but I wonder whether a `make_apply` function is better than building the record manually -- on the other hand, this change is so trivial I didn't want to bother nojb with this. @def-lkb seems interested in taking care of this PR to bring it to merge-point (but for some reason I cannot assign it to him).

nojb commented:
I changed the code to build the record manually and rebased on trunk (so that the `[@inline]` attribute is also kept in the `lambda_apply` type as well).  Also renamed the field `ap_must_tail` (which does not make grammatical sense) to the longer `ap_should_be_tailcall`.
